lib/elixir/lib/module/checker.ex

```diff
@@ -0,0 +1,130 @@
+defmodule Module.Checker do
+  def verify(module_map, _binary) do
+    state = :maps.with([:module, :file, :compile_opts], module_map)
+
+    module_map.definitions
+    |> Enum.reverse()
+    |> check_definitions(state)
+    |> List.flatten()
+  end
+
+  defp check_definitions(definitions, state) do
+    Enum.map(definitions, &check_definition(&1, state))
+  end
+
+  defp check_definition({def, _kind, meta, clauses}, state) do
```

Comment:
Nitpick: the variable `def` might be bit confusing, what do you think about `def_pair` or `def_tuple`?